OpenRA.Mods.Cnc/Traits/Disguise.cs

```diff
@@ -211,14 +223,58 @@ void HandleDisguise(Player oldEffectiveOwner, bool oldDisguiseSetting)
 
 		void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barrel)
 		{
-			if (info.RevealDisguiseOn.HasFlag(RevealDisguiseType.Attack))
+			if (info.RevealDisguiseOn.HasFlag(RevealDisguiseType.Attack) || info.RevealDisguiseOn.HasFlag(RevealDisguiseType.Demolish))
 				DisguiseAs(null);
 		}
 
 		void INotifyDamage.Damaged(Actor self, AttackInfo e)
 		{
+			if (e.Damage.Value == 0)
+				return;
+
 			if (info.RevealDisguiseOn.HasFlag(RevealDisguiseType.Damaged) && e.Damage.Value > 0)
 				DisguiseAs(null);
 		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (self.IsDisabled())
```

Comment:
We're in the process of deprecating IsDisabled so I don't think it's a good idea to introduce it in this trait.